File: src/InventorShims-vb/PathShim.vb
Imports Inventor

Public Class PathShim

    Shared Function UpOneLevel(ByVal path As String) As String

        Dim delimiter As Char

        If path.Contains(IO.Path.DirectorySeparatorChar) Then
            delimiter = IO.Path.DirectorySeparatorChar
        ElseIf path.Contains(IO.Path.AltDirectorySeparatorChar) Then
            delimiter = IO.Path.AltDirectorySeparatorChar
        Else
            'no slashes?  not a path...
            Return Nothing
        End If


        'Catch paths which end with a delimiter, such as C:\Work\Stuff\
        'Clean up so that they look like C:\Work\Stuff
        path = TrimEndingDirectorySeparator(path)

        Dim delimPos As Integer = path.LastIndexOf(delimiter)
        If (delimPos = 0) Then
            Return Nothing
        ElseIf (delimPos < 0) Then
            Return path
        End If

        Return Left(path, delimPos + 1)

    End Function

    Shared Function IsLibraryPath(ByVal path As String, ByRef inventorApp As Inventor.Application) As Boolean

        If (String.IsNullOrEmpty(path)) Then Return False

        Dim designproject As Inventor.DesignProject = inventorApp.DesignProjectManager.ActiveDesignProject
        Dim librarypaths As ProjectPaths = designproject.LibraryPaths
        Dim librarypath As ProjectPath

        For Each librarypath In librarypaths
            If path.Contains(TrimEndingDirectorySeparator(librarypath.Path)) Then
                Return True
            End If
        Next

        Return False
    End Function

    Shared Function IsContentCenterPath(ByVal path As String, ByRef inventorApp As Inventor.Application) As Boolean

        If (String.IsNullOrEmpty(path)) Then Return False

        Dim designProject As Inventor.DesignProject = inventorApp.DesignProjectManager.ActiveDesignProject
        Dim useProjectCCPath As Boolean = designProject.ContentCenterPathOverridden
        Dim ccpath As String

        If useProjectCCPath Then
            ccpath = designProject.ContentCenterPath
        Else
            ccpath = inventorApp.FileOptions.ContentCenterPath
        End If

        ccpath = TrimEndingDirectorySeparator(ccpath)

        If path.Contains(ccpath) Then
            Return True
        End If

        Return False
    End Function
    Shared Function TrimEndingDirectorySeparator(ByVal path As String) As String

        If path.EndsWith(IO.Path.DirectorySeparatorChar) Then
            path = Left(path, path.LastIndexOf(IO.Path.DirectorySeparatorChar))
        ElseIf path.EndsWith(IO.Path.AltDirectorySeparatorChar) Then
            path = Left(path, path.LastIndexOf(IO.Path.AltDirectorySeparatorChar))
        End If

        Return path

    End Function

End Class

File: src/InventorShims/PathShim.cs
using System;
using Inventor;

namespace InventorShims
{
    /// <summary>
    /// A collection of static methods to manipulate and test string paths
    /// </summary>
    public static class PathShim
    {
        /// <summary>
        /// Provided a string directory path, returns the path with the last
        /// directory removed.  E.g. UpOneLevel("C:\Work\Stuff\") returns "C:\Work"
        /// </summary>
        /// <param name="path">The path as a string</param>
        /// <returns>A string</returns>
        public static string UpOneLevel(string path)
        {
            char delimiter;
            char DirectorySeparatorChar = System.IO.Path.DirectorySeparatorChar;
            char AltDirectorySeparatorChar = System.IO.Path.AltDirectorySeparatorChar;

            if (path.Contains(DirectorySeparatorChar.ToString()))
            {
                delimiter = DirectorySeparatorChar;
            }
            else if (path.Contains(AltDirectorySeparatorChar.ToString()))
            {
                delimiter = AltDirectorySeparatorChar;
            }
            else //no slashes?  not a path...
            {
                return null;
            }

            //Catch paths which end with a delimiter, such as C:\Work\Stuff\
            //Clean up so that they look like C:\Work\Stuff
            path = TrimEndingDirectorySeparator(path);

            int delimPosition = path.LastIndexOf(delimiter);
            if (delimPosition == 0)
            {
                return null;
            }
            else if (delimPosition < 0)
            {
                return path;
            }

            return path.Remove(delimPosition + 1);
        }

        /// <summary>
        /// Provided a string that ends in a "/" or "\", returns the string without those
        /// ending characters.
        /// </summary>
        /// <param name="path">string</param>
        /// <returns>string</returns>
        public static string TrimEndingDirectorySeparator(string path)
        {
            char DirectorySeparatorChar = System.IO.Path.DirectorySeparatorChar;
            char AltDirectorySeparatorChar = System.IO.Path.AltDirectorySeparatorChar;

            if (path.EndsWith(DirectorySeparatorChar.ToString()))
            {
                path = path.Remove(path.LastIndexOf(DirectorySeparatorChar));
            }
            else if (path.EndsWith(AltDirectorySeparatorChar.ToString()))
            {
                path = path.Remove(path.LastIndexOf(AltDirectorySeparatorChar));
            }
            return path;
        }

        /// <summary>
        /// A function that checks if the provided string contains an Inventor library path.
        /// </summary>
        /// <param name="path">Path as a string.</param>
        /// <param name="inventorApp">Inventor.Application object.</param>
        /// <returns></returns>
        public static bool IsLibraryPath(this string path, Application inventorApp)
        {
            if (String.IsNullOrEmpty(path)) return false;

            DesignProject designProject = inventorApp.DesignProjectManager.ActiveDesignProject;
            ProjectPaths libraryPaths = designProject.LibraryPaths;
            //ProjectPath libraryPath;

            foreach (ProjectPath libraryPath in libraryPaths)
            {
                if (path.Contains(TrimEndingDirectorySeparator(libraryPath.Path)))
                {
                    return true;
                }
            }

            return false;
        }

        /// <summary>
        /// A function that checks if the provided string contains a Content Center Library path.
        /// </summary>
        /// <param name="path">Path as a string.</param>
        /// <param name="inventorApp">Inventor.Application object.</param>
        /// <returns></returns>
        public static bool IsContentCenterPath(this string path, Application inventorApp)
        {
            if (String.IsNullOrEmpty(path)) return false;

            DesignProject designProject = inventorApp.DesignProjectManager.ActiveDesignProject;
            bool projectCCPathInUse = designProject.ContentCenterPathOverridden;
            string ccPath;

            if (projectCCPathInUse)
            {
                ccPath = designProject.ContentCenterPath;
            } else
            {
                ccPath = inventorApp.FileOptions.ContentCenterPath;
            }

            ccPath = TrimEndingDirectorySeparator(ccPath);

            if (path.Contains(ccPath)) return true;

            return false;
        }
    }
}